contracts/token/LinearVestingWithPenalty.sol

```diff
@@ -0,0 +1,544 @@
+                tempAmounts[unclaimedCount] = releasable;
+                unclaimedCount++;
+            }
+        }
+
+        // Create result arrays with exact size of unclaimed schedules
+        indexes = new uint256[](unclaimedCount);
+        amounts = new uint256[](unclaimedCount);
+
+        // Copy collected data to result arrays
+        for (uint256 j = 0; j < unclaimedCount; j++) {
+            indexes[j] = tempIndexes[j];
+            amounts[j] = tempAmounts[j];
+        }
+
+        return (indexes, amounts);
+    }
+
+
+    /**
+    * @dev Claims the entire releasable amount for specific vesting schedule indexes with an applied penalty.
+    * @param scheduleIndexes An array of indexes of the vesting schedules to claim from.
+    */
+    function claimSpecificSchedulesWithPenalty(uint256[] calldata scheduleIndexes) external {
+        if (vestingStartDate > block.timestamp) revert VestingHasNotStartedYet();
+
+        uint256 totalNetToUser = 0;
+        uint256 totalPenaltyAmount = 0;
+
+        for (uint256 i = 0; i < scheduleIndexes.length; i++) {
+            uint256 index = scheduleIndexes[i];
+            if (index >= vestingSchedules[msg.sender].length) revert InvalidScheduleIndex();
+
+            VestingSchedule storage schedule = vestingSchedules[msg.sender][index];
+            
+            // Calculate the releasable (vested) amount that must be released without penalty
+            uint256 releasable = _releasableAmount(schedule);
+            
+            // Calculate the unvested amount
+            uint256 unvestedAmount = schedule.totalAmount - schedule.released - releasable;
+            
+            // Calculate the penalty amount on the unvested portion
+            uint256 penaltyAmount = (unvestedAmount * penaltyPercentage) / 100;
+            
+            // Calculate the net amount for the user: full releasable + (unvested - penalty)
+            uint256 netAmountToUser = releasable + (unvestedAmount - penaltyAmount);
+
+            // Accumulate amounts for total transfer
+            totalNetToUser += netAmountToUser;
+            totalPenaltyAmount += penaltyAmount;
+
+            // Mark the entire schedule as claimed by setting released to totalAmount
+            schedule.released = schedule.totalAmount;
+        }
+
+        require(totalNetToUser > 0, "No tokens available for release");
+
+        // Transfer the penalty amount to the penalty recipient, if any
+        if (totalPenaltyAmount > 0) {
+            sophtoken.safeTransfer(penaltyRecipient, totalPenaltyAmount);
+            emit PenaltyPaid(msg.sender, totalPenaltyAmount);
+        }
+
+        // Transfer the total net amount (releasable + post-penalty unvested) to the beneficiary
+        sophtoken.safeTransfer(msg.sender, totalNetToUser);
+        emit TokensReleased(msg.sender, totalNetToUser, totalPenaltyAmount);
+
+        // Burn the claimed vSOPH tokens equivalent to the total released amount plus penalties
```

Comment:
this is fine as is, but i think i would have called `_burn` before `sophtoken.safeTransfer`, feel more natural i guess

Reply:
yea easy to fix

Reply:
fixed